### ISSUE TITLE
K-means didn't converge

### ISSUE THREAD

**Issue by kh12043** [SarahRastegar/SelEx#2]
First of all, thank you for sharing the code, and I think you’ve done an amazing job. 
However, when I trained your code from scratch, I encountered an issue where center_shift=nan occurred, causing the KMeans algorithm to not converge, which led to the training not finishing. Below is a log from a specific epoch where the process was running for over 8 hours.

`[running kmeans]: 19588it [8:21:09, 1.54s/it, center_shift=nan, iteration=1958
`
As for the hyperparameters, I only modified `--grad_from_block = 10` for the CUB dataset and left everything else unchanged. Additionally, the same issue occurred when I used the Stanford Cars dataset.

Is there something I might have done wrong?

I would appreciate your guidance on this issue. Looking forward to your response, and thanks again for your hard work!

**Comment by SarahRastegar** [comment]
Thanks for your interest in our work! Could you clarify whether you’re using the PyTorch KMeans or scikit-learn KMeans implementation? If the process is getting stuck in an epoch, it could indicate an issue either with the PyTorch KMeans version or with the initialization of the cluster centers.

If you’re using my adapted version of SSKMeans, you might try switching to the traditional SSKMeans implementation to see if that resolves the issue. If it does, then the problem is likely related to the initialization of the centers. In that case, you can try disabling the first initialization step in the adapted SSKMeans or reducing the cluster size (it doesn't need to be exact for the initialization).

If you’re still encountering issues, especially if you see NaN values, it suggests something isn't working correctly. Unfortunately, without more information, I can't pinpoint the exact cause, but these suggestions should help avoid the process getting stuck. Let me know if any of these solutions work for you!

**Comment by kh12043** [comment]
I used PyTorch KMeans, and the error mentioned above occurred. When I used the KMeans provided by Scikit-learn, the error did not occur. 

However, the experimental results on the Stanford Cars dataset are somewhat different from those in the paper. (All: 56.12 | Old: 75.81 | New: 46.61)
Could it be that I misconfigured something in the settings, such as the hyperparameters? (unsupervised_smoothing=0.5, grad_from_block=10, the rest of the hyperparameters are set to the provided defaults) 
Thank you for your response!

**Comment by SarahRastegar** [comment]
Hi! Could you please share the type of GPU you're using?

The Pytorch KMeans implementation was not only faster but also provided significantly better clustering results. Since our method relies on these cluster assignments as pseudolabels, higher clustering accuracy leads to better overall performance in a ripple effect.

Given that the rest of the code is working without Pytorch KMeans, it seems that the core setup is fine. To resolve this issue, please try installing KMeans Pytorch directly in the main directory by running the following commands:

```
cd SelEx
git clone https://github.com/subhadarship/kmeans_pytorch
cd kmeans_pytorch
pip install --editable .
```

Let me know if that helps or if you encounter any issues during the installation (also remove any kmeans_pytorch installation you had before).

**Comment by kh12043** [comment]
I reinstalled kmeans_pytorch as you instructed and ran the experiment again, but the results on the Stanford Cars dataset are not much different from what I reported earlier. Could I have missed something?

**Comment by SarahRastegar** [comment]
Have you set the hyperparameters as mentioned in the README? Please look at this [issue](https://github.com/SarahRastegar/SelEx/issues/6), I suggested some notes to reproduce the papers' results. Please let me know if this fixed the problem.
